src/xenia/apu/xma_context.cc

```diff
@@ -243,162 +346,319 @@ void XmaContext::DecodePackets(XMA_CONTEXT_DATA* data) {
+      if (data->loop_count > 0 &&
+          data->input_buffer_read_offset == data->loop_end) {
+        // Loop back to the beginning.
+        data->input_buffer_read_offset = data->loop_start;
+        if (data->loop_count < 255) {
+          data->loop_count--;
+        }
+      } else if (!partial && len > 0) {
+        data->input_buffer_read_offset += len;
+      }
+    } else if (len < 0) {
+      // Did not get frame
+      XELOGAPU("libav failed to decode a frame!");
+      if (frame_size && frame_size != 0x7FFF) {
+        data->input_buffer_read_offset += frame_size;
+      } else {
         data->input_buffer_0_valid = 0;
         data->input_buffer_1_valid = 0;
       }
-    } else {
-      // Decoder is out of data and there's no more to give.
-      break;
+      return;
     }
-  }
 
-  // The game will kick us again with a new output buffer later.
-  data->output_buffer_valid = 0;
-}
+    last_input_read_pos_ = data->input_buffer_read_offset;
 
-int XmaContext::StartPacket(XMA_CONTEXT_DATA* data) {
-  // Translate pointers for future use.
-  uint8_t* in0 = data->input_buffer_0_valid
-                     ? memory()->TranslatePhysical(data->input_buffer_0_ptr)
-                     : nullptr;
-  uint8_t* in1 = data->input_buffer_1_valid
-                     ? memory()->TranslatePhysical(data->input_buffer_1_ptr)
-                     : nullptr;
+    if (got_frame) {
+      // Successfully decoded a frame.
+      // Copy to the output buffer.
+      size_t written_bytes = 0;
 
-  int sample_rate = GetSampleRate(data->sample_rate);
-  int channels = data->is_stereo ? 2 : 1;
-
-  // See if we've finished with the input.
-  // Block count is in packets, so expand by packet size.
-  uint32_t input_size_0_bytes = data->input_buffer_0_valid
-                                    ? (data->input_buffer_0_packet_count) * 2048
-                                    : 0;
-  uint32_t input_size_1_bytes = data->input_buffer_1_valid
-                                    ? (data->input_buffer_1_packet_count) * 2048
-                                    : 0;
-
-  // Total input size
-  uint32_t input_size_bytes = input_size_0_bytes + input_size_1_bytes;
-
-  // Input read offset is in bits. Typically starts at 32 (4 bytes).
-  // "Sequence" offset - used internally for WMA Pro decoder.
-  // Just the read offset.
-  uint32_t seq_offset_bytes = (data->input_buffer_read_offset & ~0x7FF) / 8;
-  uint32_t input_remaining_bytes = input_size_bytes - seq_offset_bytes;
-
-  if (seq_offset_bytes < input_size_bytes) {
-    // Setup input offset and input buffer.
-    uint32_t input_offset_bytes = seq_offset_bytes;
-    auto input_buffer = in0;
-
-    if (seq_offset_bytes >= input_size_0_bytes && input_size_1_bytes) {
-      // Size overlap, select input buffer 1.
-      // TODO(DrChat): This needs testing.
-      input_offset_bytes -= input_size_0_bytes;
-      input_buffer = in1;
-    }
+#ifdef DEBUG
```

Comment:
If it's useful as an assert you could just change them to that (like the buffer size check?) - otherwise kill em :)